File: Data Structures & Algorithms/course-schedule-ii/submission-1.cpp
class Solution {
    unordered_map<int, vector<int>> preMap;

    unordered_set<int> visiting;
    unordered_set<int> visited;

    vector<int> result;
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        for (int i = 0; i < numCourses; i++) {
            preMap[i] = {}; // make sure every course is in the map
        }
        for (auto &p : prerequisites) {
            preMap[p[1]].push_back(p[0]);
        }

        // run dfs for each course
        for (int crs = 0; crs < numCourses; crs++) {
            if (!dfs(crs)) return {}; // cycle found -> empty result
        }

        // reverse because dfs builds in postorder
        reverse(result.begin(), result.end());
        return result;
    }


    bool dfs(int crs){
        if (visiting.count(crs)) return false; // cycle
        if (visited.count(crs)) return true;   // already processed


        visiting.insert(crs);

        for (int pre : preMap[crs]){
            if(!dfs(pre))return false;
        }

        visiting.erase(crs);
        visited.insert(crs);
        result.push_back(crs);

    return true;
    }
};

File: Data Structures & Algorithms/course-schedule/submission-1.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        preMap = {}  # Empty dictionary

        for i in range(numCourses):
            preMap[i] = []  # Initialize each course with empty list
            print(preMap)

        for crs, pre in prerequisites:
            preMap[crs].append(pre)
       

            # Store all courses along the current DFS path
        visiting = set()

        def dfs(crs):
            if crs in visiting:
                return False
            if preMap[crs] == []:
                return True

            visiting.add(crs)

            for pre in preMap[crs]:
                if not dfs(pre):
                    return False
            visiting.remove(crs)
            preMap[crs] = []
            return True


        for c in range(numCourses):
            if not dfs(c):
                return False
        return True


